### DISCUSSION TITLE
Build root not found - Standard Configuration?

### DISCUSSION THREAD

**Discussion by DentOpt** [nasa/fprime#1625] [Q&A] [answered]
| | |
|:---|:---|
|**_F´ Version _**| v3.1.0|
|**_cmake_**|  3.16.3 |
---
We are migrating our software to Fprime v2.1.0 to v3.1.0 with the following structure:
```
~/workspace
 - Fprime
 - MyProj
```
(build folder in MyProj)
However, with the new fprime version "No build root found".
What parametrization am I missing for v3.1.0? 

Error message:
```
-- Performing CMake source prescan
CMake Error at ~/workspace/Fprime/cmake/utilities.cmake:410 (message):
  No build root found for: ~/workspace/MyProj
Call Stack (most recent call first):
  ~/workspace/Fprime/cmake/utilities.cmake:445 (get_nearest_build_root)
  ~/workspace/Fprime/cmake/API.cmake:309 (get_module_name)
  CMakeLists.txt:53 (register_fprime_deployment)


CMake Error at ~/workspace/Fprime/cmake/prescan.cmake:91 (message):
-- Configuring incomplete, errors occurred!
  Failed to run prescan, cannot continue.
See also "~/ws_fprime/workspace/MyProj/build-fprime-automatic-native/CMakeFiles/CMakeOutput.log".
Call Stack (most recent call first):
See also "~/ws_fprime/workspace/MyProj/build-fprime-automatic-native/CMakeFiles/CMakeError.log".
  ~/workspace/Fprime/cmake/FPrime.cmake:68 (perform_prescan)
  CMakeLists.txt:32 (include)

[ERROR] CMake erred with return code 1. Partial build cache remains. Run purge to clean-up.
```

## How to Reproduce
```
mkdir workspace
cd workspace
git clone https://github.com/nasa/fprime.git Fprime --branch v3.1.0
cp -r Fprime/Ref MyProj
```

In `MyProj/settings.ini`:
```
[fprime]
project_root: .
framework_path: ../Fprime
```

In `MyProj/CmakeLists.txt`: substitute
```
include("${CMAKE_CURRENT_LIST_DIR}/../cmake/FPrime.cmake")
# NOTE: register custom targets between these two lines
include("${CMAKE_CURRENT_LIST_DIR}/../cmake/FPrime-Code.cmake")\
```
with
```
include("${CMAKE_CURRENT_LIST_DIR}/../Fprime/cmake/FPrime.cmake")
# NOTE: register custom targets between these two lines
include("${CMAKE_CURRENT_LIST_DIR}/../Fprime/cmake/FPrime-Code.cmake")
```

Build 
```
cd MyProj
fprime-util generate # Error is thrown 
```

## Expected Behavior

`fprime-util generate` passes without error

**Comment by LeStarch** [comment, marked as answer]
I think you need to set `project_root` to `..`.

`project_root` represents the root directory for all modules in your F´  project.  It is relative to `settings.ini` and since `settings.ini` is in `MyProj` using `.` as I see above means it cannot locate `MyProj` within `MyProj` and produces the error you see.  Changing to `..` means it will look for `MyProj` under ~/workspace.


Technically speaking F´ looks for some root containing `MyProj` and those roots are: "fprime_location" for F´  modules, "library_locations" for modules coming from a library, and "project_root" for project supplied modules. None of these locations contain `MyProj` as the `project_root` is set to the folder `MyProj` not its parent folder. Hence "no root found" error is caused because no location searched has a child folder `MyProj`.

In your case, the logical root for your item is the project_root, controlled by the project_root setting, which is why we focused there and noticed the error.

**Reply by LeStarch** [reply]
Also, thanks for the clear explanation of the problem.  It made it fairly straight-forward to solve.

**Reply by DentOpt** [reply]
Thank you for your quick answer! - That solves the issue. 

From my perspective, it would be more intuitive/flexibility to be able to treat Fprime as isolated external dependency whereas the user code should be the root of the development. Is there a strong reason for the current setup?

**Reply by LeStarch** [reply]
I understand this is a bit of a quirk with F´, part of it is historical, part of it was decided based on the below information.  However, I welcome further questions, comments, and suggestions such that we can improve!

F´ can be treated as an external entity.  You could download and install F´  in `/opt/fprime` and set `framework_path` to `/opt/fprime` for example.

The trick is knowing where user-code or project-code "starts".  Is the root of your project `~/`?  `~/workspace`?  `~/workspace/MyProj`?  In the case of Git, you can detect a `.git` folder, but that may be inaccurate, and we don't strictly require projects to manage code in `git` although almost all do.  Thus we decided to have the user just set the path in the `settings.ini`.

I also think you are asking why project code goes under `MyProj` instead of living at the root level of the project.  All components are fully-named in the build system by the path from the "root".  Thus, MyProj/Gnc/Imu has the name MyProj_Gnc_Imu. Often it is useful to include the `MyProj` token at the front to prevent target collisions.  Some projects ship with multiple deployments (e.g. `MyProjRemoteNode` and `MyProjHostNode`.  These projects want the flexibility to have multiple `settings.ini` that are independent within a single project structure.

Again, I welcome thoughts and suggestions here.  This structure came from the days when F´ and project code were together (e.g. fprime/Ref) and we needed a way to separate concerns.  Now that F´ allows separated project code, there may be better practices that could be adopted.

**Reply by DentOpt** [reply]
Thanks for your explanation. I agree that git is not the best solution, as you might manage the modules via submodules. 
I think it is a good approach to set the root, (for me it would makes sense that the root and naming typically starts from the "MyProj" root). 
All the paths should be defined relative to this root, so that there are no constraint where the other packages sit.
